### DISCUSSION TITLE
Introduction of GeoVista at PyVista tutorial in SciPy2023

### DISCUSSION THREAD

**Discussion by tkoyama010** [bjlittle/geovista#343] [Ideas]
@bjlittle

I will be working on the [7. PyVista in Action section of the pyvista tutorial](https://github.com/pyvista/pyvista-tutorial#tutorial-outline) in SciPy2023. I plan to spend a lot of time explaining GeoVista in this section.

If there are any GeoVista features or benefits you would like to see covered in the tutorial, please send us your comments. Suggested examples are also welcome.

Edit: The example here is particularly touching.

[🎥 WW3 SMC time-series](https://github.com/bjlittle/geovista/assets/2051656/876d877e-a6fa-42ff-8153-08c41ff8a19e)

**Comment by bjlittle** [comment]
Hey @tkoyama010,

This is amazing, thanks so much!

Sadly, I can't join you at the `pyvista` tutorial as I don't arrive at SciPy until late on Monday evening 😞.  But I'd love to help support you as best as I can.

How can I easily do that? What do you need/want? 

I've already banked quite a few examples, see `geovista.examples`, which showcase various initial use-cases. All the examples have associated data to make them executable, which `geovista` will download automatically. Also see the CLI i.e., `geovista --help` where you can easily download any/all `geovista` assets (`geovista download --all`) or run any/all of the examples (`geovista examples --all --verbose`).

Note that, `geovista` uses [pooch](https://github.com/fatiando/pooch) to manage downloads of versioned assets.

Also see, `geovista.samples` (which uses `geovista.pantry`) which underpins the `geovista.examples`, and has a stock of meshes ready for you to visualise out-the-box and play with.

Obviously, one major feature is projection support, which allows you to convert your geo-located mesh in cartesian space to a target CRS (Coordinate Reference System). At the moment, only basic cylindrical and pseudo-cylindrical projections are supported, but as `geovista` matures a fuller treatment of those CRS will be available, and also other classes of projections, such as azimuthal and conic et al.

At the moment, I'm currently working on supporting coastline projections, so that should be available this week in `geovista` 0.3.0, along with gridline (graticule) support and `kd-tree` searching support.

We've also hinted at how the power of `pyvista` and `vtk` can be leveraged to extract mesh regions without using `shapely` i.e., see `geovista.geodesic`, in particular `geovista.geodesic.BBox`, `geovista.geodesic.panel`, `geovista.geodesic.wedge` and `geovista.geodesic.line`.

Essentially, further primitives will be provided in the near future that will allow users to easily extract regions from their data using caps, bands, shapefiles or other geometries e.g., extract the State of Texas, or the UK or whatever random polygon the user cares about.

In general, I'd say the following about `geovista`:

- It's early days for `geovista`, but I think the potential is kinda clear for people to see; for me the potential is huge
- I'm keen that `geovista` doesn't reinvent the wheel and simply "wrap" `pyvista` functionality. I want `geovista` to be a clean and simple gateway into the world of `pyvista`, but with cartographic conveniences
- I've put quite a bit of effort into `geovista.bridge`, which allows users to convert their uniform, rectilinear, and curvilinear grids and unstructured meshes to `pyvista`. So that's an attractive point to perhaps labour to your audience. We also recently added point-cloud support, see `geovista.bridge.Transform.from_points`
- I'm super excited about [pyvista-xarray](https://github.com/pyvista/pyvista-xarray) and the potential of `geovista` underpinning some of it's capability, but also I'm very keen to see the `xarray` community rally behind that project, simply because @banesullivan is going to do amazing things! I mean, a paradigm shift game changer, and that's pretty exciting... well, it is to me at least 😉 

Let me know what you think... but I'm sure I can give you some additional material to support your introductory talk on `geovista` 👍

**Reply by tkoyama010** [reply]
Thank you for your comment. Your reply gave me a lot of inspiration.

I would like to script a tutorial focusing on your recommendations `geovista.geodesic` and `geovista.bridge`. If you have time for us before the tutorial takes place, we appreciate it if you can review it.

Focusing on `pyvista-xarray` is a very good point of view! I will add an explanation about it in the tutorial. Can you tell us more about how the GeoVista project plans to utilize `pyvista-xarray` in the future?

**Reply by bjlittle** [reply]
@tkoyama010 Here's something to get you started...

At the Met Office we are moving to an unstructured cube-sphere mesh... which is a cube projected out onto a sphere i.e., there are six panels on the sphere. Each cube-sphere is defined by the number of "cells squared" within each panel e.g., the following example is a C48 cube-sphere, so there are `6 * 48 * 48` cells.

```python
import geovista.samples
import geovista.theme

c48 = geovista.samples.lfric(resolution="c48")
c48.plot(show_edges=True)
```
![image](https://github.com/bjlittle/geovista/assets/2051656/6e0da9eb-a44e-47ce-a5e3-5bcc800d6263)

Here's a sample C48 cube-sphere populated with Sea Surface Temperature data (on the cells):

```python
c48_sst = geovista.samples.lfric_sst()
c48_sst.plot(show_edges=True)
```

![image](https://github.com/bjlittle/geovista/assets/2051656/6feb77c8-65e8-49b4-a578-c8df10639553)

Note that, the land masses are masked, as oceanographers, like fish, typically aren't interested in land (a sweeping generalisation 😉)

There is a convenience within `geovista.geodesic` that creates a `geovista.geodesic.BBox` instance for any 1 of the 6 cube-sphere panels i.e., `geovista.geodesic.panel`

```python
from geovista.geodesic import panel

bbox = panel("americas")
bbox.mesh.plot()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/5067ba48-8291-453f-bd34-ab34df889854)

Note that, this bounding box (`bbox`) is constructed from geodesic lines i.e., great circles, and is a 3D manifold. As such, we can then use it to extract points/cells from any underlying mesh. Before doing that, first let's render the bounding box and the mesh together so that we can see their relationship to one another. Note that, our `bbox` instance is indeed covering the correct panel of the cube-sphere.

```python
import geovista as gv

plotter = gv.GeoPlotter()
plotter.add_mesh(c48_sst, show_edges=True)
plotter.add_mesh(bbox.mesh)
plotter.add_axes()
plotter.view_yz()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/75a62140-b803-448d-9fb7-c4c7bdba7f3d)

As a fun exercise, you could play with `opacity` on the `bbox.mesh` to see through the manifold to the underlying cube-sphere surface, or turn on the gridlines of the bbox etc

We can also see the footprint of the `bbox` boundary on the surface of the target mesh, which is perhaps a wee bit clearer to some users:

```python
plotter = gv.GeoPlotter()
plotter.add_mesh(c48_sst, show_edges=True)
plotter.add_mesh(bbox.boundary(), color="green", line_width=5)
plotter.add_axes()
plotter.view_xz()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/ba829012-7d7a-4378-83d0-1555b73b8bb2)

Let's now use the bounding box to extract the mesh that it encloses:

```python
region = bbox.enclosed(c48_sst)

plotter = gv.GeoPlotter()
plotter.add_mesh(region, show_edges=True)
plotter.add_axes()
plotter.view_xz()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/6f0a0cf5-368a-41ae-bd70-f1c2e113cebd)

You could perhaps then play with the `preference` kwarg of the `bbox.enclosed` method to see the impact on the end result.

However, let's geo-locate the `region` by also rendering a texture mapped base layer in addition to some coastlines:

```python
plotter = gv.GeoPlotter()
plotter.add_mesh(region, show_edges=True)
plotter.add_coastlines()
plotter.add_base_layer(texture=gv.natural_earth_hypsometric())
plotter.view_xz()
plotter.show_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/bb3ab687-9e32-4deb-90f2-074592561148)

Also, as we're not so interested in the land mask, let's threshold that out and re-spin the render:

```python
sea_region = region.threshold()

plotter = gv.GeoPlotter()
plotter.add_mesh(sea_region, show_edges=True)
plotter.add_coastlines()
plotter.add_base_layer(texture=gv.natural_earth_hypsometric())
plotter.view_xz()
plotter.show_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/f852375d-31dd-457c-997b-e7862b81d7d8)

Since we're here... let's transform the `sea_region` to a Robinson projection:

```python
plotter = gv.GeoPlotter(crs="+proj=robin lon_0=-90")
plotter.add_mesh(sea_region, show_edges=True)
plotter.add_coastlines()
plotter.add_base_layer(texture=gv.natural_earth_hypsometric())
plotter.view_xy()
plotter.show_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/e326eb2e-8413-44b7-9fd3-1673358c3bf5)

It's also easily possible to get the inverted result i.e., the surface of the mesh not enclosed by the `bbox` manifold:

```python
outside = bbox.enclosed(c48_sst, outside=True)

plotter = gv.GeoPlotter()
plotter.add_mesh(outside, show_edges=True)
plotter.add_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/68f5e909-0a48-44c6-99b1-f9a7b548a16c)

Akin to before, let's render this again, but with a base layer underneath:

```python
plotter = gv.GeoPlotter()
plotter.add_mesh(outside, show_edges=True)
plotter.add_coastlines()
plotter.add_base_layer(texture=gv.natural_earth_1())
plotter.view_xz()
plotter.show_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/3f1aac15-d771-4d44-b9a1-36b0bc037025)

It's not quite clear what's going on here, although playing with the render interactively helps, but let's transform the mesh to the Mollweider projection to help clarify matters:

```python
plotter = gv.GeoPlotter(crs="+proj=moll lon_0=-90")
plotter.add_mesh(outside, show_edges=True)
plotter.add_coastlines()
plotter.add_base_layer(texture=gv.natural_earth_1())
plotter.view_xy()
plotter.show_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/7718a6e3-359b-4942-a773-cc0b10fecfd7)

And again, let's remove the land mask so that we can see more of the texture mapped base layer:

```python
sea_outside = outside.threshold()

plotter = gv.GeoPlotter(crs="+proj=moll lon_0=-90")
plotter.add_mesh(sea_outside, show_edges=True)
plotter.add_coastlines()
plotter.add_base_layer(texture=gv.natural_earth_1())
plotter.view_xy()
plotter.show_axes()
plotter.show()
```

![image](https://github.com/bjlittle/geovista/assets/2051656/4270b0fd-0eb5-4ada-8aab-eef04939007c)

Sweet! 😄 

----

Food for thought perhaps?

Also, explore the `BBox` class to create custom bounding box instances, and there is also the `geovista.geodesic.wedge`, a convenience function akin to the `geovista.geodesic.panel`. Plus you can easily render geodesic lines i.e., great circles, with `geovista.geodesic.line`.

The point here is that this is just the first step. I'm aiming to provide a richer suite of such primitives to extract regions in similar ways. But the capability showcased by `geovista.geodesic` hints at the direction of where I'm taking `geovista`. The other point to make is that thanks to `pyvista` and `vtk` the extraction operation is pretty darn fast as opposed to other traditional approaches (perhaps I should garner metrics to back that up!)

Any good?

**Reply by bjlittle** [reply]
@tkoyama010 BTW totally happy to review and contribute to any material that you propose 👍

**Reply by tkoyama010** [reply]
@bjlittle This is great. thanks a lot! I will introduce it in a tutorial. I'll ask participants to add stars to GeoVista 😉.

**Reply by bjlittle** [reply]
@tkoyama010 Hey! I've updated the example above with some extra steps that gives it a bit more flow. Let me know what you think 

Also, I'll be releasing `geovista` 0.3.0 next week, so I'll ping you when that happens and you can update your `pyvista` tutorial dependency for `geovista` 👍

**Reply by tkoyama010** [reply]
Thanks a lot. I added your example in this [PR](https://github.com/pyvista/pyvista-tutorial/pull/136). It will be a great tutorial!

**Reply by tkoyama010** [reply]
@bjlittle
I will try to explain what each package ties to pyvista when I explain the tutorial. (e.g. sunkit-pyvista is a bridge between sunpy and pyvista, pyvista-xarray is pyvista and xarray, etc.)
Which package would you say geovista trying makes bridge to pyvista?
(My current understanding is that it is weather data using netCDF4, but I would like to hear from the author if this is correct.)

**Reply by bjlittle** [reply]
@tkoyama010 I think clarifying the relationship between third-party packages and `pyvista` would be really valuable to the audience. It helps give them essential context :100:

In this regard, I think `geovista` is slightly different. I'd say that `geovista` is a "gateway" to `pyvista` rather than a "bridge". As in, it's attempting to be agnostic to any other third-party package such as `xarray` or `sunpy` (or indeed `iris`). Yes, my examples primarily use `netCDF4`, but that's simply as a convenience to load some data of interest. There would be nothing to stop someone with an `xarray` dataset or a `sunpy` dataset using `geovista`... although they'd have to do a wee bit more work to prepare their data, as they don't have the convenience that `pyvista-xarray` or `sunkit-pyvista` offer of rendering natively "out-the-box".

So, I think that `geovista` is at a slighly lower level than `pyvista-xarray` et al i.e., more generic and less specific, as in, not tied to a particular third-party package.

I guess my point here is that, if you present the `geovista.bridge` with a [NEP-18](https://numpy.org/neps/nep-0018-array-function-protocol.html) compliant array, then it should "just work". The reason that I've taken this approach is that it opens up `geovista` to being used by anyone in the community, whereas `pyvista-xarray` is specifically targeting the `xarray` community, and `sunpy-pyvista` is specifically targeting the `sunpy` community. And both of those packages will offer streamlined workflows for those specific communities, which is completely priceless and of huge benefit.

Does this all make sense?

**Reply by tkoyama010** [reply]
@bjlittle 
That is what you said in [motivation in README](https://github.com/bjlittle/geovista#motivation)! 
I notice that there is a tool developed by gateway concept as well as a bridge concept. You broaden my horizons!

**Reply by tkoyama010** [reply]
The explanation that GeoVista is on the shoulders of PyVista and Numpy, just as Cartopy is on the shoulders of giants in the Matplotlib and Numpy data formats, is easy to understand.

**Reply by bjlittle** [reply]
@tkoyama010 Perfect. That is indeed the message in a nutshell! ❤️

**Reply by tkoyama010** [reply]
And I found that GeoVista is [Iris’ Mesh Partner Packages](https://scitools-iris.readthedocs.io/en/stable/further_topics/ugrid/partner_packages.html#iris-mesh-partner-packages).

**Reply by bjlittle** [reply]
@tkoyama010 Yes, that's correct. I'm a core developer on SciTools/iris, have been for the last 10 years or so.

Essentially, `geovista` was borne out or work that we done on `iris` to extend it's data model to support unstructured meshes, which now has an implementation of the [UGRID Convention](http://ugrid-conventions.github.io/ugrid-conventions/). We also developed unstructured support for regridding through [iris-esmf-regrid](https://github.com/SciTools-incubator/iris-esmf-regrid). 

So, `geovista` is all part of that toolkit. We'll shortly be integrating `geovista` with `iris`, akin to the goal of `pyvista-xarray`, so that `iris` data cubes can be easily rendered using `geovista` (note that, the glue code for this interface will not live in `geovista`, which will remain agnostic to `iris`, as mentioned previously)

**Reply by tkoyama010** [reply]
I was suspicious as to why GeoVista provides `GeoPlotter`. But, I noticed that `GeoPlotter` has earth science-specific methods such as `add_base_layer` and `add_coastlines`. I realized that this is indeed the "gateway" you are talking about. 
(In my understanding, "GeoVista translating abstract PyVista concepts into more concrete geoscience concepts".)

For a beginner in PyVista tutorial, using `GeoPlotter` a priori will hinder understanding, so I will change our explanation approach to use PyVista's `Plotter` first to plot and explain the usefulness of `GeoPlotter` a posteriori.

**Reply by tkoyama010** [reply]
> So, geovista is all part of that toolkit. We'll shortly be integrating geovista with iris, akin to the goal of pyvista-xarray, so that iris data cubes can be easily rendered using geovista (note that, the glue code for this interface will not live in geovista, which will remain agnostic to iris, as mentioned previously)

@bjlittle This is a good interview with background on the development of GeoVista. Thanks!

**Reply by bjlittle** [reply]
@tkoyama010 Yup, the `GeoPlotter` is specifically adding cartographic behaviours for users who want to render maps and geolocated data.

So the `GeoPlotter` is a specialisation of the `pyvista.Plotter` i.e., it has extra cartographic bells and whistles

**Reply by tkoyama010** [reply]
@bjlittle
I have a question about the tutorial you created, about LFRic, is [this article](https://www.metoffice.gov.uk/research/approach/modelling-systems/lfric) the most reliable information?
My specialty is the use of finite element methods in earthquake engineering, so I am impressed by the attention to the use of finite element and finite difference methods in the field of meteorology as well.

**Reply by tkoyama010** [reply]
I found the same link in [README](https://github.com/bjlittle/geovista#lfric-cube-sphere). I don't need to ask if I read the document more. Very sorry.

**Reply by tkoyama010** [reply]
Since there are still 3 minutes left in the tutorial time, I will explain how the `Transform.from_2d` method allows users to define their own longitude, latitude, and data with NumPy's `masked_array` to GeoVista.
Participants will be able to experience the possibility of connecting their `numpy` and `xarray` geographic data to GeoVista (and to PyVista)!

**Reply by bjlittle** [reply]
> @bjlittle
> I have a question about the tutorial you created, about LFRic, is [this article](https://www.metoffice.gov.uk/research/approach/modelling-systems/lfric) the most reliable information?
> My specialty is the use of finite element methods in earthquake engineering, so I am impressed by the attention to the use of finite element and finite difference methods in the field of meteorology as well.

@tkoyama010 You've done your research! I'm impressed!

I guess that's a reasonable public facing article to reference. It's a bit old now, but still highly relevant 👍

**Reply by bjlittle** [reply]
> Since there are still 3 minutes left in the tutorial time, I will explain how the `Transform.from_2d` method allows users to define their own longitude, latitude, and data with NumPy's `masked_array` to GeoVista.
> Participants will be able to experience the possibility of connecting their `numpy` and `xarray` geographic data to GeoVista (and to PyVista)!

@tkoyama010 Brilliant! That's a great idea. It sounds like the tutorial is shaping up to be really fun. I hope it all goes well.

There are plenty of examples (and associated data) that come with `geovista` if people don't have data but want to play around. Take a peek in `geovista.examples` for executable examples, where `geovista` will automatically download and cache the required data. 

The `geovista.examples` get their data from the `geovista.pantry`, but there are also "ready-to-go" meshes built from the `geovista.pantry` data in `geovista.samples`.

Hopefully, you'll have enough data there to keep people busy 🥳 

BTW I've just release `geovista` 0.3.0, so please update the dependencies of the tutorial 😀

**Comment by bjlittle** [comment]
> Focusing on `pyvista-xarray` is a very good point of view! I will add an explanation about it in the tutorial. Can you tell us more about how the GeoVista project plans to utilize `pyvista-xarray` in the future?

I think the relationship is really the other way around. `geovista` won't be utilizing `pyvista-xarray`, it will be `pyvista-xarray` that will be utilizing `geovista`. Perhaps have a chat with @banesullivan to get a feel for his vision and roadmap of where he wants to take `pyvista-xarray`. But the potential is huge, given the size of the Earth Sciences `xarray` community.

**Comment by bjlittle** [comment]
@tkoyama010 Are you okay with me closing this discussion now?
